### DISCUSSION TITLE
Option to parallelize chains across workers?

### DISCUSSION THREAD

**Discussion by stuvet** [wlandau/stantargets#46] [Ideas]
As I understand it, the following runs 4 chains on a single machine, each with 4 threads - 16vcpu machine in total, or 32vcpu if not using hyperthreading.

```
tar_stan_mcmc(
     ...,
     chains = 4,
     parallel_chains = 4,
     threads_per_chain = 4,
     ...
)
```

Occasionally this will leave 24vcpus of a 32vcpu machine sitting idle for a few hours while a single slow chain finishes.

When using transient workers via `future.batchtools` it'd be better to release these unnecessary vcpus.

Ideally it'd be submitted as an array job, but it looks like future.batchtools [doesn't support this](https://github.com/HenrikBengtsson/future/issues/256)?

A nice-to-have workaround would be for `tar_stan_mcmc` to have an option which would internally parallelize the mcmc target to run single chains across 4 targets then combine them again before running the summary, draws & diagnostics jobs.

Can you suggest any workaround within stantargets?

Thanks

**Comment by wlandau** [comment]
I see what you are saying, but at the moment I still prefer to keep all the chains in a single target. Advantages:

* Close alignment with the `cmstanr` R package interface. The arguments you mention are directly forwarded to `cmdstanr` functions.
* Consistent behavior across `tar_stan_mcmc()` vs `tar_stan_mcmc_rep_summary()` and other already dynamically branched targets.
* Avoid unmanageably large numbers of targets in large Bayesian data analysis pipelines: https://books.ropensci.org/targets/practices.html#how-much-a-target-should-do

Perhaps this is an issue to raise with the `cmdstanr` and/or CmdStan maintainers because these same resource consumption issues are shared by `cmdstanr`/CmdStan users that do not use `stantargets`. When one CmdStan chain finishes, it would ideally release the vcpus while the other chains keep running on other vcpus.

**Reply by stuvet** [reply]
I think cmdstan does release the vcpus, but I believe a e.g. a Slurm cluster controller doesn't reallocate resources on nodes until a job is done.

As far as I know future.batchtools doesn't play nicely with job arrays (which would spread the 4 chains across multiple machines).

I'm no expert but I have a sneaky feeling a different slurm config will help here - I think I can tell it to treat cpus as resources rather than just the nodes. I'll update this thread once I've tested it.

**Reply by stuvet** [reply]
I thought wrong. I'm getting no luck changing the Slurm config, even with the builtin scheduler, SelectType=select/cons_tres, --oversubscribe, --spread-jobs, --nodes=4 etc

I suspect cmdstanr is launching a single process, which then launches the 4 chains, so the Slurm controller is allocating them all to a single machine. 

I suspect the only way to get around it would be to launch single chain jobs then merge them afterwards, but as you said this is likely better done by cmdstanr. I'll take a look at their code & have a word over there if it looks like this is what's going on.

Thanks.
